src/object/srv_obj_migrate.c

```diff
@@ -3853,49 +3934,302 @@ ds_obj_migrate_handler(crt_rpc_t *rpc)
 	dss_rpc_reply(rpc, DAOS_REBUILD_DROP_OBJ);
 }
 
+static int
+obj_tree_lookup_cont(daos_handle_t toh, uuid_t co_uuid, daos_handle_t *cont_toh)
+{
+	struct tree_cache_root	*cont_root = NULL;
+	d_iov_t			key_iov;
+	d_iov_t			tmp_iov;
+	int			rc;
+
+	d_iov_set(&key_iov, co_uuid, sizeof(uuid_t));
+	d_iov_set(&tmp_iov, NULL, 0);
+	rc = dbtree_lookup(toh, &key_iov, &tmp_iov);
+	if (rc < 0) {
+		if (rc != -DER_NONEXIST)
+			D_ERROR("lookup cont "DF_UUID" failed, "DF_RC"\n",
+				DP_UUID(co_uuid), DP_RC(rc));
+		else
+			D_DEBUG(DB_TRACE, "Container "DF_UUID" not exist\n",
+				DP_UUID(co_uuid));
+		return rc;
+	}
+
+	cont_root = tmp_iov.iov_buf;
+	*cont_toh = cont_root->tcr_root_hdl;
+	return 0;
+}
+
+static int
+obj_tree_lookup_uoid(daos_handle_t cont_toh, daos_unit_oid_t uoid)
+{
+	struct migrate_obj_val	val;
+	d_iov_t			key_iov;
+	d_iov_t			val_iov;
+
+	d_iov_set(&key_iov, &uoid, sizeof(uoid));
+	d_iov_set(&val_iov, &val, sizeof(struct migrate_obj_val));
+	return dbtree_lookup(cont_toh, &key_iov, &val_iov);
+}
+
+#define REINT_ITER_YIELD_CNT	(256)
+struct reint_post_iter_arg {
+	daos_handle_t			 ria_migrated_tree_hdl;
+	struct migrate_pool_tls		*ria_tls;
+	daos_handle_t			 ria_cont_toh;
+	uuid_t				 ria_co_uuid;
+	int				 ria_yield_cnt;
+};
+
+static int
+reint_post_obj_iter_cb(daos_handle_t ch, vos_iter_entry_t *entry,
+		       vos_iter_type_t type, vos_iter_param_t *iter_param,
+		       void *data, unsigned *acts)
+{
+	struct reint_post_iter_arg	*arg = data;
+	struct migrate_pool_tls		*tls = arg->ria_tls;
+	daos_unit_oid_t			 uoid = entry->ie_oid;
+	int				 rc = 0;
+
+	/* for non-replica obj don't try to delete it */
+	if (daos_obj_id2ord(uoid.id_pub) == OR_RP_1)
+		goto out;
```

Comment:
How to handle the following case: a 2-way object, related servers on which the two shards reside are both offline for some reason almost at the same time, then nobody update such object after that, and then when we reint the two servers, we should keep such 2-way object, right? Does current logic support that? Here, we need to discard such object, right?

Reply:
right, this is a known issue to me. I wanted to discuss with you at weekly meeting but did not get time.
One explanation is with RP2 obj then RF <= 1, then losing 2 engines cannot ensure no data lose. But we should try the best to not lost data.
Anyway, we cannot handle everything in one patch, let's discuss and resolve it in following PR.